// @twurple/api/es/api/helix/stream/HelixStreamApi.mjs
var HelixStreamApi_1;
import { __decorate } from "tslib";
import { flatten } from '@d-fischer/shared-utils';
import { HttpStatusCodeError } from '@twurple/api-call';
import { extractUserId, extractUserName, rtfm } from '@twurple/common';
import { StreamNotLiveError } from "../../../errors/StreamNotLiveError.mjs";
import { BaseApi } from "../../BaseApi.mjs";
import { HelixPaginatedRequest } from "../HelixPaginatedRequest.mjs";
import { createPaginatedResult } from "../HelixPaginatedResult.mjs";
import { makePaginationQuery } from "../HelixPagination.mjs";
import { HelixTag } from "../tag/HelixTag.mjs";
import { HelixStream } from "./HelixStream.mjs";
import { HelixStreamMarker } from "./HelixStreamMarker.mjs";
import { HelixStreamMarkerWithVideo } from "./HelixStreamMarkerWithVideo.mjs";
/**
 * The Helix API methods that deal with streams.
 *
 * Can be accessed using `client.streams` on an {@ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const stream = await api.streams.getStreamByUserId('125328655');
 * ```
 */
let HelixStreamApi = HelixStreamApi_1 = class HelixStreamApi extends BaseApi {
    /**
     * Retrieves a list of streams.
     *
     * @expandParams
     */
    async getStreams(filter = {}) {
        const result = await this._client.callApi({
            url: 'streams',
            type: 'helix',
            query: {
                ...makePaginationQuery(filter),
                community_id: filter.community,
                game_id: filter.game,
                language: filter.language,
                type: filter.type,
                user_id: filter.userId,
                user_login: filter.userName
            }
        });
        return createPaginatedResult(result, HelixStream, this._client);
    }
    /**
     * Creates a paginator for streams.
     *
     * @expandParams
     */
    getStreamsPaginated(filter = {}) {
        return new HelixPaginatedRequest({
            url: 'streams',
            query: {
                community_id: filter.community,
                game_id: filter.game,
                language: filter.language,
                type: filter.type,
                user_id: filter.userId,
                user_login: filter.userName
            }
        }, this._client, data => new HelixStream(data, this._client));
    }
    /**
     * Retrieves the current stream for the given user name.
     *
     * @param user The user name to retrieve the stream for.
     */
    async getStreamByUserName(user) {
        const result = await this.getStreams({ userName: extractUserName(user) });
        return result.data.length ? result.data[0] : null;
    }
    /**
     * Retrieves the current stream for the given user ID.
     *
     * @param user The user ID to retrieve the stream for.
     */
    async getStreamByUserId(user) {
        const result = await this.getStreams({ userId: extractUserId(user) });
        return result.data.length ? result.data[0] : null;
    }
    /**
     * Retrieves a list of all stream markers for an user.
     *
     * @param user The user to list the stream markers for.
     * @param pagination
     *
     * @expandParams
     */
    async getStreamMarkersForUser(user, pagination) {
        return await this._getStreamMarkers('user_id', extractUserId(user), pagination);
    }
    /**
     * Creates a paginator for all stream markers for an user.
     *
     * @param user The user to list the stream markers for.
     */
    getStreamMarkersForUserPaginated(user) {
        return this._getStreamMarkersPaginated('user_id', extractUserId(user));
    }
    /**
     * Retrieves a list of all stream markers for a video.
     *
     * @param videoId The video to list the stream markers for.
     * @param pagination
     *
     * @expandParams
     */
    async getStreamMarkersForVideo(videoId, pagination) {
        return await this._getStreamMarkers('video_id', videoId, pagination);
    }
    /**
     * Creates a paginator for all stream markers for a video.
     *
     * @param videoId The video to list the stream markers for.
     */
    getStreamMarkersForVideoPaginated(videoId) {
        return this._getStreamMarkersPaginated('video_id', videoId);
    }
    /**
     * Creates a new stream marker.
     *
     * Only works while the specified user's stream is live.
     *
     * @param broadcaster The broadcaster to create a stream marker for.
     * @param description The description of the marker.
     */
    async createStreamMarker(broadcaster, description) {
        try {
            const result = await this._client.callApi({
                url: 'streams/markers',
                method: 'POST',
                type: 'helix',
                scope: 'user:edit:broadcast',
                query: {
                    user_id: extractUserId(broadcaster),
                    description
                }
            });
            return new HelixStreamMarker(result.data[0], this._client);
        }
        catch (e) {
            if (e instanceof HttpStatusCodeError && e.statusCode === 404) {
                throw new StreamNotLiveError();
            }
            throw e;
        }
    }
    /**
     * Retrieves the tags of a stream.
     *
     * @param broadcaster The broadcaster of the stream.
     */
    async getStreamTags(broadcaster) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'streams/tags',
            query: {
                broadcaster_id: extractUserId(broadcaster)
            }
        });
        return result.data.map(data => new HelixTag(data));
    }
    /**
     * Replaces the tags of a stream.
     *
     * @param broadcaster The broadcaster of the stream.
     * @param tagIds The tags to set. If not given, removes all tags.
     */
    async replaceStreamTags(broadcaster, tagIds) {
        await this._client.callApi({
            type: 'helix',
            url: 'streams/tags',
            scope: 'user:edit:broadcast',
            method: 'PUT',
            query: {
                broadcaster_id: extractUserId(broadcaster)
            },
            jsonBody: {
                tag_ids: tagIds
            }
        });
    }
    /**
     * Retrieves the stream key of a stream.
     *
     * @param broadcaster The broadcaster to retrieve the stream key for.
     */
    async getStreamKey(broadcaster) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'streams/key',
            scope: 'channel:read:stream_key',
            query: {
                broadcaster_id: extractUserId(broadcaster)
            }
        });
        return result.data[0].stream_key;
    }
    /**
     * Retrieves the streams that are currently live and are followed by the given user.
     *
     * @param user The user to check followed streams for.
     * @param pagination
     *
     * @expandParams
     */
    async getFollowedStreams(user, pagination) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'streams/followed',
            scope: 'user:read:follows',
            query: {
                user_id: extractUserId(user),
                ...makePaginationQuery(pagination)
            }
        });
        return createPaginatedResult(result, HelixStream, this._client);
    }
    /**
     * Creates a paginator for the streams that are currently live and are followed by the given user.
     *
     * @param user The user to check followed streams for.
     */
    getFollowedStreamsPaginated(user) {
        return new HelixPaginatedRequest({
            url: 'stream/followed',
            scope: 'user:read:follows',
            query: {
                user_id: extractUserId(user)
            }
        }, this._client, data => new HelixStream(data, this._client));
    }
    async _getStreamMarkers(queryType, id, pagination) {
        var _a;
        const result = await this._client.callApi({
            url: 'streams/markers',
            type: 'helix',
            query: {
                [queryType]: id,
                ...makePaginationQuery(pagination)
            },
            scope: 'user:read:broadcast'
        });
        return {
            data: flatten(result.data.map(HelixStreamApi_1._mapGetStreamMarkersResult.bind(this._client))),
            cursor: (_a = result.pagination) === null || _a === void 0 ? void 0 : _a.cursor
        };
    }
    _getStreamMarkersPaginated(queryType, id) {
        return new HelixPaginatedRequest({
            url: 'streams/markers',
            query: {
                [queryType]: id
            },
            scope: 'user:read:broadcast'
        }, this._client, HelixStreamApi_1._mapGetStreamMarkersResult.bind(this._client));
    }
    static _mapGetStreamMarkersResult(data) {
        return data.videos.reduce((result, video) => [
            ...result,
            ...video.markers.map(marker => new HelixStreamMarkerWithVideo(marker, video.video_id, this))
        ], []);
    }
};
HelixStreamApi = HelixStreamApi_1 = __decorate([
    rtfm('api', 'HelixStreamApi')
], HelixStreamApi);
export { HelixStreamApi };
