// @twurple/eventsub/lib/events/EventSubChannelModeratorEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSubChannelModeratorEvent = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * An EventSub event representing a broadcaster adding or removing a moderator in their channel.
 */
let EventSubChannelModeratorEvent = class EventSubChannelModeratorEvent extends common_1.DataObject {
    /** @private */
    constructor(data, client) {
        super(data);
        this._client = client;
    }
    /**
     * The ID of the broadcaster that added or removed a moderator.
     */
    get broadcasterId() {
        return this[common_1.rawDataSymbol].broadcaster_user_id;
    }
    /**
     * The name of the broadcaster that added or removed a moderator.
     */
    get broadcasterName() {
        return this[common_1.rawDataSymbol].broadcaster_user_login;
    }
    /**
     * The display name of the broadcaster that added or removed a moderator.
     */
    get broadcasterDisplayName() {
        return this[common_1.rawDataSymbol].broadcaster_user_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.users.getUserById(this[common_1.rawDataSymbol].broadcaster_user_id));
    }
    /**
     * The ID of the user that was added or removed as a moderator.
     */
    get userId() {
        return this[common_1.rawDataSymbol].user_id;
    }
    /**
     * The name of the user that was added or removed as a moderator.
     */
    get userName() {
        return this[common_1.rawDataSymbol].user_login;
    }
    /**
     * The display name of the user that was added or removed as a moderator.
     */
    get userDisplayName() {
        return this[common_1.rawDataSymbol].user_name;
    }
    /**
     * Retrieves more information about the user.
     */
    async getUser() {
        return (await this._client.users.getUserById(this[common_1.rawDataSymbol].user_id));
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], EventSubChannelModeratorEvent.prototype, "_client", void 0);
EventSubChannelModeratorEvent = tslib_1.__decorate([
    common_1.rtfm('eventsub', 'EventSubChannelModeratorEvent', 'userId')
], EventSubChannelModeratorEvent);
exports.EventSubChannelModeratorEvent = EventSubChannelModeratorEvent;
