// @twurple/eventsub/lib/events/EventSubStreamOnlineEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSubStreamOnlineEvent = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * An EventSub event representing a stream going live.
 */
let EventSubStreamOnlineEvent = class EventSubStreamOnlineEvent extends common_1.DataObject {
    /** @private */
    constructor(data, client) {
        super(data);
        this._client = client;
    }
    /**
     * The ID of the broadcaster.
     */
    get broadcasterId() {
        return this[common_1.rawDataSymbol].broadcaster_user_id;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this[common_1.rawDataSymbol].broadcaster_user_login;
    }
    /**
     * The display name of the broadcaster.
     */
    get broadcasterDisplayName() {
        return this[common_1.rawDataSymbol].broadcaster_user_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.users.getUserById(this[common_1.rawDataSymbol].broadcaster_user_id));
    }
    /**
     * Retrieves more information about the stream.
     */
    async getStream() {
        return (await this._client.streams.getStreamByUserId(this[common_1.rawDataSymbol].broadcaster_user_id));
    }
    /**
     * The type of the stream going live.
     */
    get streamType() {
        return this[common_1.rawDataSymbol].type;
    }
    /**
     * The date and time when the stream was started.
     */
    get startDate() {
        return new Date(this[common_1.rawDataSymbol].started_at);
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], EventSubStreamOnlineEvent.prototype, "_client", void 0);
EventSubStreamOnlineEvent = tslib_1.__decorate([
    common_1.rtfm('eventsub', 'EventSubStreamOnlineEvent', 'broadcasterId')
], EventSubStreamOnlineEvent);
exports.EventSubStreamOnlineEvent = EventSubStreamOnlineEvent;
