// @twurple/api/es/api/helix/game/HelixGameApi.mjs
import { __decorate } from "tslib";
import { rtfm } from '@twurple/common';
import { BaseApi } from "../../BaseApi.mjs";
import { HelixPaginatedRequest } from "../HelixPaginatedRequest.mjs";
import { createPaginatedResult } from "../HelixPaginatedResult.mjs";
import { makePaginationQuery } from "../HelixPagination.mjs";
import { HelixGame } from "./HelixGame.mjs";
/**
 * The Helix API methods that deal with games.
 *
 * Can be accessed using `client.games` on an {@ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const game = await api.games.getGameByName('Hearthstone');
 * ```
 */
let HelixGameApi = class HelixGameApi extends BaseApi {
    /**
     * Retrieves the game data for the given list of game IDs.
     *
     * @param ids The game IDs you want to look up.
     */
    async getGamesByIds(ids) {
        return await this._getGames('id', ids);
    }
    /**
     * Retrieves the game data for the given list of game names.
     *
     * @param names The game names you want to look up.
     */
    async getGamesByNames(names) {
        return await this._getGames('name', names);
    }
    /**
     * Retrieves the game data for the given game ID.
     *
     * @param id The game ID you want to look up.
     */
    async getGameById(id) {
        const games = await this._getGames('id', [id]);
        return games.length ? games[0] : null;
    }
    /**
     * Retrieves the game data for the given game name.
     *
     * @param name The game name you want to look up.
     */
    async getGameByName(name) {
        const games = await this._getGames('name', [name]);
        return games.length ? games[0] : null;
    }
    /**
     * Retrieves a list of the most viewed games at the moment.
     *
     * @param pagination
     *
     * @expandParams
     */
    async getTopGames(pagination) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'games/top',
            query: makePaginationQuery(pagination)
        });
        return createPaginatedResult(result, HelixGame, this._client);
    }
    /**
     * Creates a paginator for the most viewed games at the moment.
     */
    getTopGamesPaginated() {
        return new HelixPaginatedRequest({
            url: 'games/top'
        }, this._client, data => new HelixGame(data, this._client));
    }
    async _getGames(filterType, filterValues) {
        if (!filterValues.length) {
            return [];
        }
        const result = await this._client.callApi({
            type: 'helix',
            url: 'games',
            query: {
                [filterType]: filterValues
            }
        });
        return result.data.map(entry => new HelixGame(entry, this._client));
    }
};
HelixGameApi = __decorate([
    rtfm('api', 'HelixGameApi')
], HelixGameApi);
export { HelixGameApi };
