// @twurple/eventsub/lib/events/common/EventSubChannelHypeTrainContribution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSubChannelHypeTrainContribution = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
class EventSubChannelHypeTrainContribution extends common_1.DataObject {
    /** @private */
    constructor(data, client) {
        super(data);
        this._client = client;
    }
    /**
     * The contributor's ID.
     */
    get userId() {
        return this[common_1.rawDataSymbol].user_id;
    }
    /**
     * The contributor's user name.
     */
    get userName() {
        return this[common_1.rawDataSymbol].user_login;
    }
    /**
     * The contributor's display name.
     */
    get userDisplayName() {
        return this[common_1.rawDataSymbol].user_name;
    }
    /**
     * Retrieves more information about the contributor.
     */
    async getUser() {
        return (await this._client.users.getUserById(this[common_1.rawDataSymbol].user_id));
    }
    /**
     * The type of the contribution.
     */
    get type() {
        return this[common_1.rawDataSymbol].type;
    }
    /**
     * The contributor's total contribution.
     */
    get total() {
        return this[common_1.rawDataSymbol].total;
    }
}
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], EventSubChannelHypeTrainContribution.prototype, "_client", void 0);
exports.EventSubChannelHypeTrainContribution = EventSubChannelHypeTrainContribution;
