// httpanda/lib/parseUrl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseUrl = exports.parseQueryString = void 0;
var tslib_1 = require("tslib");
function parseQueryString(str) {
    var e_1, _a;
    var result = {};
    try {
        for (var _b = tslib_1.__values(str.split('&')), _c = _b.next(); !_c.done; _c = _b.next()) {
            var part = _c.value;
            var eqPos = part.indexOf('=');
            var key = '', value = part;
            if (eqPos !== -1) {
                key = part.substr(0, eqPos);
                value = part.substr(eqPos + 1);
            }
            result[key] = Object.prototype.hasOwnProperty.call(result, key)
                ? Array.isArray(result[key])
                    ? tslib_1.__spread(result[key], [value]) : [result[key], value]
                : value;
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return result;
}
exports.parseQueryString = parseQueryString;
function parseUrl(req) {
    var _a;
    var url = req.url;
    if (url === undefined) {
        url = '/';
    }
    if (((_a = req._parseUrlCache) === null || _a === void 0 ? void 0 : _a._full) === url) {
        return req._parseUrlCache;
    }
    var searchStart = url.indexOf('?');
    if (searchStart === -1) {
        return (req._parseUrlCache = {
            href: url,
            path: url,
            pathname: url,
            search: null,
            query: {},
            _full: url
        });
    }
    var search = url.substr(searchStart);
    return (req._parseUrlCache = {
        href: url,
        path: url,
        pathname: url.substr(0, searchStart),
        search: search,
        query: parseQueryString(search.substr(1)),
        _full: url
    });
}
exports.parseUrl = parseUrl;
