// @twurple/eventsub/lib/EventSubMiddleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSubMiddleware = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@twurple/common");
const checks_1 = require("./checks");
const EventSubBase_1 = require("./EventSubBase");
/**
 * An Express middleware for the Twitch EventSub event distribution mechanism.
 *
 * You can find an extensive example on how to use this class in the [documentation](/docs/getting-data/eventsub/express).
 *
 * @hideProtected
 * @inheritDoc
 */
let EventSubMiddleware = class EventSubMiddleware extends EventSubBase_1.EventSubBase {
    /**
     * Creates a new EventSub middleware wrapper.
     *
     * @param config
     *
     * @expandParams
     */
    constructor(config) {
        super(config);
        checks_1.checkHostName(config.hostName);
        this._hostName = config.hostName;
        this._pathPrefix = config.pathPrefix;
    }
    /**
     * Applies middleware that handles EventSub notifications to an Express app.
     *
     * @param app The app the middleware should be applied to.
     */
    async apply(app) {
        let pathPrefix = this._pathPrefix;
        if (pathPrefix) {
            pathPrefix = `/${pathPrefix.replace(/^\/|\/$/g, '')}`;
        }
        const requestHandler = this._createHandleRequest();
        const dropLegacyHandler = this._createDropLegacyRequest();
        const healthHandler = this._createHandleHealthRequest();
        if (pathPrefix) {
            app.post(`${pathPrefix}/event/:id`, requestHandler);
            app.post(`${pathPrefix}/:id`, dropLegacyHandler);
            app.get(`${pathPrefix}`, healthHandler);
        }
        else {
            app.post('event/:id', requestHandler);
            app.post(':id', dropLegacyHandler);
            app.get('/', healthHandler);
        }
    }
    /**
     * Marks the middleware as ready to receive events.
     *
     * The express app should be started before this.
     */
    async markAsReady() {
        this._readyToSubscribe = true;
        await this._resumeExistingSubscriptions();
    }
    async getHostName() {
        return this._hostName;
    }
    async getPathPrefix() {
        return this._pathPrefix;
    }
};
EventSubMiddleware = tslib_1.__decorate([
    common_1.rtfm('eventsub', 'EventSubMiddleware')
], EventSubMiddleware);
exports.EventSubMiddleware = EventSubMiddleware;
