// @twurple/api/lib/api/badges/ChatBadgeVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatBadgeVersion = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@twurple/common");
/**
 * A version of a badge.
 */
let ChatBadgeVersion = class ChatBadgeVersion extends common_1.DataObject {
    /**
     * The action to execute when the badge is clicked.
     */
    get clickAction() {
        return this[common_1.rawDataSymbol].click_action;
    }
    /**
     * The URL to visit when the badge is clicked.
     *
     * Only applies if clickAction === 'visit_url'.
     */
    get clickUrl() {
        return this[common_1.rawDataSymbol].click_url;
    }
    /**
     * The description of the badge.
     */
    get description() {
        return this[common_1.rawDataSymbol].description;
    }
    /**
     * Gets an image URL for the given scale.
     *
     * @param scale The scale of the badge image.
     */
    getImageUrl(scale) {
        return this[common_1.rawDataSymbol][`image_url_${scale}x`];
    }
    /**
     * The title of the badge.
     */
    get title() {
        return this[common_1.rawDataSymbol].title;
    }
};
ChatBadgeVersion = tslib_1.__decorate([
    common_1.rtfm('api', 'ChatBadgeVersion')
], ChatBadgeVersion);
exports.ChatBadgeVersion = ChatBadgeVersion;
