// @twurple/api/es/api/helix/moderation/HelixModeratorEvent.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { DataObject, rawDataSymbol, rtfm } from '@twurple/common';
/**
 * An event that indicates the change of a moderator status, i.e. gaining or losing moderation privileges.
 */
let HelixModeratorEvent = class HelixModeratorEvent extends DataObject {
    /** @private */
    constructor(eventData, client) {
        super(eventData);
        this._client = client;
    }
    /**
     * The unique ID of the moderator event.
     */
    get eventId() {
        return this[rawDataSymbol].id;
    }
    /**
     * The type of the moderator event.
     */
    get eventType() {
        return this[rawDataSymbol].event_type;
    }
    /**
     * The date of the moderator event.
     */
    get eventDate() {
        return new Date(this[rawDataSymbol].event_timestamp);
    }
    /**
     * The version of the moderator event.
     */
    get eventVersion() {
        return this[rawDataSymbol].version;
    }
    /**
     * The ID of the user.
     */
    get userId() {
        return this[rawDataSymbol].event_data.user_id;
    }
    /**
     * The name of the user.
     */
    get userName() {
        return this[rawDataSymbol].event_data.user_login;
    }
    /**
     * The display name of the user.
     */
    get userDisplayName() {
        return this[rawDataSymbol].event_data.user_name;
    }
    /**
     * Retrieves more information about the user.
     */
    async getUser() {
        return (await this._client.users.getUserById(this[rawDataSymbol].event_data.user_id));
    }
    /**
     * The id of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterId() {
        return this[rawDataSymbol].event_data.broadcaster_id;
    }
    /**
     * The name of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterName() {
        return this[rawDataSymbol].event_data.broadcaster_login;
    }
    /**
     * The display name of the broadcaster where the user gained/lost moderation privileges.
     */
    get broadcasterDisplayName() {
        return this[rawDataSymbol].event_data.broadcaster_name;
    }
    /**
     * Retrieves more information about the broadcaster.
     */
    async getBroadcaster() {
        return (await this._client.users.getUserById(this[rawDataSymbol].event_data.broadcaster_id));
    }
};
__decorate([
    Enumerable(false)
], HelixModeratorEvent.prototype, "_client", void 0);
HelixModeratorEvent = __decorate([
    rtfm('api', 'HelixModeratorEvent', 'userId')
], HelixModeratorEvent);
export { HelixModeratorEvent };
