// @twurple/api/es/api/helix/moderation/HelixModerationApi.mjs
import { __decorate } from "tslib";
import { extractUserId, rtfm } from '@twurple/common';
import { BaseApi } from "../../BaseApi.mjs";
import { HelixPaginatedRequest } from "../HelixPaginatedRequest.mjs";
import { createPaginatedResult } from "../HelixPaginatedResult.mjs";
import { makePaginationQuery } from "../HelixPagination.mjs";
import { HelixBan } from "./HelixBan.mjs";
import { HelixBanEvent } from "./HelixBanEvent.mjs";
import { HelixModerator } from "./HelixModerator.mjs";
import { HelixModeratorEvent } from "./HelixModeratorEvent.mjs";
/**
 * The Helix API methods that deal with moderation.
 *
 * Can be accessed using `client.moderation` on an {@ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const { data: users } = await api.moderation.getBannedUsers('61369223');
 * ```
 */
let HelixModerationApi = class HelixModerationApi extends BaseApi {
    /**
     * Retrieves a list of banned users in a given channel.
     *
     * @param channel The channel to retrieve the banned users from.
     * @param filter Additional filters for the result set.
     */
    async getBannedUsers(channel, filter) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/banned',
            scope: 'moderation:read',
            query: {
                broadcaster_id: extractUserId(channel),
                user_id: filter === null || filter === void 0 ? void 0 : filter.userId,
                ...makePaginationQuery(filter)
            }
        });
        return createPaginatedResult(result, HelixBan, this._client);
    }
    /**
     * Creates a paginator for banned users in a given channel.
     *
     * @param channel The channel to retrieve the banned users from.
     */
    getBannedUsersPaginated(channel) {
        return new HelixPaginatedRequest({
            url: 'moderation/banned',
            scope: 'moderation:read',
            query: {
                broadcaster_id: extractUserId(channel)
            }
        }, this._client, data => new HelixBan(data, this._client), 50 // possibly a relatively consistent workaround for twitchdev/issues#18
        );
    }
    /**
     * Checks whether a given user is banned in a given channel.
     *
     * @param channel The channel to check for a ban of the given user.
     * @param user The user to check for a ban in the given channel.
     */
    async checkUserBan(channel, user) {
        const userId = extractUserId(user);
        const result = await this.getBannedUsers(channel, { userId });
        return result.data.some(ban => ban.userId === userId);
    }
    /**
     * Retrieves a list of ban events for a given channel.
     *
     * @param channel The channel to retrieve the ban events from.
     * @param filter Additional filters for the result set.
     */
    async getBanEvents(channel, filter) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/banned/events',
            scope: 'moderation:read',
            query: {
                broadcaster_id: extractUserId(channel),
                user_id: filter === null || filter === void 0 ? void 0 : filter.userId,
                ...makePaginationQuery(filter)
            }
        });
        return createPaginatedResult(result, HelixBanEvent, this._client);
    }
    /**
     * Creates a paginator for ban events for a given channel.
     *
     * @param channel The channel to retrieve the ban events from.
     */
    getBanEventsPaginated(channel) {
        return new HelixPaginatedRequest({
            url: 'moderation/banned/events',
            scope: 'moderation:read',
            query: {
                broadcaster_id: extractUserId(channel)
            }
        }, this._client, data => new HelixBanEvent(data, this._client));
    }
    /**
     * Retrieves a list of moderators in a given channel.
     *
     * @param channel The channel to retrieve moderators from.
     * @param filter Additional filters for the result set.
     */
    async getModerators(channel, filter) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/moderators',
            scope: 'moderation:read',
            query: {
                broadcaster_id: extractUserId(channel),
                user_id: filter === null || filter === void 0 ? void 0 : filter.userId,
                ...makePaginationQuery(filter)
            }
        });
        return createPaginatedResult(result, HelixModerator, this._client);
    }
    /**
     * Creates a paginator for moderators in a given channel.
     *
     * @param channel The channel to retrieve moderators from.
     */
    getModeratorsPaginated(channel) {
        return new HelixPaginatedRequest({
            url: 'moderation/moderators',
            scope: 'moderation:read',
            query: {
                broadcaster_id: extractUserId(channel)
            }
        }, this._client, data => new HelixModerator(data, this._client));
    }
    /**
     * Checks whether a given user is a moderator of a given channel.
     *
     * @param channel The channel to check.
     * @param user The user to check.
     */
    async checkUserMod(channel, user) {
        const userId = extractUserId(user);
        const result = await this.getModerators(channel, { userId });
        return result.data.some(mod => mod.userId === userId);
    }
    /**
     * Retrieves a list of moderator events for a given channel.
     *
     * @param channel The channel to retrieve the moderator events from.
     * @param filter Additional filters for the result set.
     */
    async getModeratorEvents(channel, filter) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'moderation/moderators/events',
            scope: 'moderation:read',
            query: {
                broadcaster_id: extractUserId(channel),
                user_id: filter === null || filter === void 0 ? void 0 : filter.userId,
                ...makePaginationQuery(filter)
            }
        });
        return createPaginatedResult(result, HelixModeratorEvent, this._client);
    }
    /**
     * Creates a paginator for moderator events for a given channel.
     *
     * @param channel The channel to retrieve the moderator events from.
     */
    getModeratorEventsPaginated(channel) {
        return new HelixPaginatedRequest({
            url: 'moderation/moderators/events',
            scope: 'moderation:read',
            query: {
                broadcaster_id: extractUserId(channel)
            }
        }, this._client, data => new HelixModeratorEvent(data, this._client));
    }
    /**
     * Processes a message held by AutoMod.
     *
     * @param user The user who is processing the message.
     * @param msgId The ID of the message.
     * @param allow Whether to allow the message - `true` allows, and `false` denies.
     */
    async processHeldAutoModMessage(user, msgId, allow) {
        await this._client.callApi({
            type: 'helix',
            url: 'moderation/automod/message',
            method: 'POST',
            scope: 'moderator:manage:automod',
            jsonBody: {
                user_id: extractUserId(user),
                msg_id: msgId,
                action: allow ? 'ALLOW' : 'DENY'
            }
        });
    }
};
HelixModerationApi = __decorate([
    rtfm('api', 'HelixModerationApi')
], HelixModerationApi);
export { HelixModerationApi };
