// @twurple/api/es/api/helix/HelixApiGroup.mjs
import { __decorate } from "tslib";
import { Cacheable, CachedGetter } from '@d-fischer/cache-decorators';
import { rtfm } from '@twurple/common';
import { BaseApi } from "../BaseApi.mjs";
import { HelixBitsApi } from "./bits/HelixBitsApi.mjs";
import { HelixChannelApi } from "./channel/HelixChannelApi.mjs";
import { HelixChannelPointsApi } from "./channelPoints/HelixChannelPointsApi.mjs";
import { HelixChatApi } from "./chat/HelixChatApi.mjs";
import { HelixClipApi } from "./clip/HelixClipApi.mjs";
import { HelixEventSubApi } from "./eventSub/HelixEventSubApi.mjs";
import { HelixExtensionsApi } from "./extensions/HelixExtensionsApi.mjs";
import { HelixGameApi } from "./game/HelixGameApi.mjs";
import { HelixHypeTrainApi } from "./hypeTrain/HelixHypeTrainApi.mjs";
import { HelixModerationApi } from "./moderation/HelixModerationApi.mjs";
import { HelixPollApi } from "./poll/HelixPollApi.mjs";
import { HelixPredictionApi } from "./prediction/HelixPredictionApi.mjs";
import { HelixScheduleApi } from "./schedule/HelixScheduleApi.mjs";
import { HelixSearchApi } from "./search/HelixSearchApi.mjs";
import { HelixStreamApi } from "./stream/HelixStreamApi.mjs";
import { HelixSubscriptionApi } from "./subscriptions/HelixSubscriptionApi.mjs";
import { HelixTagApi } from "./tag/HelixTagApi.mjs";
import { HelixTeamApi } from "./team/HelixTeamApi.mjs";
import { HelixUserApi } from "./user/HelixUserApi.mjs";
import { HelixVideoApi } from "./video/HelixVideoApi.mjs";
/**
 * Groups all API calls available in Helix a.k.a. the "New Twitch API".
 *
 * Can be accessed using {@ApiClient#helix}.
 *
 * @deprecated Please remove `.helix` from your calls to access the API namespaces directly.
 */
let HelixApiGroup = class HelixApiGroup extends BaseApi {
    /**
     * The Helix bits API methods.
     */
    get bits() {
        return new HelixBitsApi(this._client);
    }
    /**
     * The Helix channels API methods.
     */
    get channels() {
        return new HelixChannelApi(this._client);
    }
    /**
     * The Helix channel points API methods.
     */
    get channelPoints() {
        return new HelixChannelPointsApi(this._client);
    }
    /**
     * The Helix chat API methods.
     */
    get chat() {
        return new HelixChatApi(this._client);
    }
    /**
     * The Helix clips API methods.
     */
    get clips() {
        return new HelixClipApi(this._client);
    }
    /**
     * The Helix EventSub API methods.
     */
    get eventSub() {
        return new HelixEventSubApi(this._client);
    }
    /**
     * The Helix extensions API methods.
     */
    get extensions() {
        return new HelixExtensionsApi(this._client);
    }
    /**
     * The Helix game API methods.
     */
    get games() {
        return new HelixGameApi(this._client);
    }
    /**
     * The Helix Hype Train API methods.
     */
    get hypeTrain() {
        return new HelixHypeTrainApi(this._client);
    }
    /**
     * The Helix moderation API methods.
     */
    get moderation() {
        return new HelixModerationApi(this._client);
    }
    /**
     * The Helix poll API methods.
     */
    get polls() {
        return new HelixPollApi(this._client);
    }
    /**
     * The Helix prediction API methods.
     */
    get predictions() {
        return new HelixPredictionApi(this._client);
    }
    /**
     * The Helix schedule API methods.
     */
    get schedule() {
        return new HelixScheduleApi(this._client);
    }
    /**
     * The Helix search API methods.
     */
    get search() {
        return new HelixSearchApi(this._client);
    }
    /**
     * The Helix stream API methods.
     */
    get streams() {
        return new HelixStreamApi(this._client);
    }
    /**
     * The Helix subscription API methods.
     */
    get subscriptions() {
        return new HelixSubscriptionApi(this._client);
    }
    /**
     * The Helix tag API methods.
     */
    get tags() {
        return new HelixTagApi(this._client);
    }
    /**
     * The Helix team API methods.
     */
    get teams() {
        return new HelixTeamApi(this._client);
    }
    /**
     * The Helix user API methods.
     */
    get users() {
        return new HelixUserApi(this._client);
    }
    /**
     * The Helix video API methods.
     */
    get videos() {
        return new HelixVideoApi(this._client);
    }
};
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "bits", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "channels", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "channelPoints", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "chat", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "clips", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "eventSub", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "extensions", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "games", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "hypeTrain", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "moderation", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "polls", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "predictions", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "schedule", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "search", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "streams", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "subscriptions", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "tags", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "teams", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "users", null);
__decorate([
    CachedGetter()
], HelixApiGroup.prototype, "videos", null);
HelixApiGroup = __decorate([
    Cacheable,
    rtfm('api', 'HelixApiGroup')
], HelixApiGroup);
export { HelixApiGroup };
