// @twurple/api/es/api/helix/bits/HelixBitsApi.mjs
import { __decorate } from "tslib";
import { extractUserId, rtfm } from '@twurple/common';
import { BaseApi } from "../../BaseApi.mjs";
import { HelixBitsLeaderboard } from "./HelixBitsLeaderboard.mjs";
import { HelixCheermoteList } from "./HelixCheermoteList.mjs";
/**
 * The Helix API methods that deal with bits.
 *
 * Can be accessed using `client.bits` on an {@ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const leaderboard = await api.bits.getLeaderboard({ period: 'day' });
 * ```
 */
let HelixBitsApi = class HelixBitsApi extends BaseApi {
    /**
     * Retrieves a bits leaderboard of your channel.
     *
     * @expandParams
     */
    async getLeaderboard(params = {}) {
        const { count = 10, period = 'all', startDate, contextUserId } = params;
        const result = await this._client.callApi({
            type: 'helix',
            url: 'bits/leaderboard',
            scope: 'bits:read',
            query: {
                count: count.toString(),
                period,
                started_at: startDate ? startDate.toISOString() : undefined,
                user_id: contextUserId
            }
        });
        return new HelixBitsLeaderboard(result, this._client);
    }
    /**
     * Retrieves all available cheermotes.
     *
     * @param broadcaster The broadcaster to include custom cheermotes of.
     *
     * If not given, only retrieves global cheermotes.
     */
    async getCheermotes(broadcaster) {
        const broadcasterId = broadcaster ? extractUserId(broadcaster) : undefined;
        const result = await this._client.callApi({
            type: 'helix',
            url: 'bits/cheermotes',
            query: {
                broadcaster_id: broadcasterId
            }
        });
        return new HelixCheermoteList(result.data);
    }
};
HelixBitsApi = __decorate([
    rtfm('api', 'HelixBitsApi')
], HelixBitsApi);
export { HelixBitsApi };
