// @twurple/api/es/api/helix/tag/HelixTagApi.mjs
import { __decorate } from "tslib";
import { rtfm } from '@twurple/common';
import { BaseApi } from "../../BaseApi.mjs";
import { HelixPaginatedRequest } from "../HelixPaginatedRequest.mjs";
import { createPaginatedResult } from "../HelixPaginatedResult.mjs";
import { makePaginationQuery } from "../HelixPagination.mjs";
import { HelixTag } from "./HelixTag.mjs";
/**
 * The Helix API methods that deal with tags.
 *
 * Can be accessed using `client.tags` on an {@ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const tags = await api.tags.getAllStreamTags();
 * ```
 */
let HelixTagApi = class HelixTagApi extends BaseApi {
    /**
     * Retrieves all stream tags.
     *
     * @param pagination
     *
     * @expandParams
     */
    async getAllStreamTags(pagination) {
        const result = await this._client.callApi({
            type: 'helix',
            url: 'tags/streams',
            query: makePaginationQuery(pagination)
        });
        return createPaginatedResult(result, HelixTag);
    }
    /**
     * Creates a paginator for all stream tags.
     */
    getAllStreamTagsPaginated() {
        return new HelixPaginatedRequest({
            url: 'tags/streams'
        }, this._client, data => new HelixTag(data));
    }
    /**
     * Retrieves a set of stream tags by IDs.
     *
     * @param ids The IDs of the stream tags.
     */
    async getStreamTagsByIds(ids) {
        if (!ids.length) {
            return [];
        }
        const result = await this._client.callApi({
            type: 'helix',
            url: 'tags/streams',
            query: {
                tag_id: ids
            }
        });
        return result.data.map(data => new HelixTag(data));
    }
    /**
     * Retrieves a single stream tag by ID.
     *
     * @param id The ID of the stream tag.
     */
    async getStreamTagById(id) {
        var _a;
        const tags = await this.getStreamTagsByIds([id]);
        return (_a = tags[0]) !== null && _a !== void 0 ? _a : null;
    }
};
HelixTagApi = __decorate([
    rtfm('api', 'HelixTagApi')
], HelixTagApi);
export { HelixTagApi };
