// @twurple/api/lib/api/helix/moderation/HelixBan.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixBan = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const common_1 = require("@twurple/common");
/**
 * Information about the ban of a user.
 */
let HelixBan = class HelixBan extends common_1.DataObject {
    /** @private */
    constructor(data, client) {
        super(data);
        this._client = client;
    }
    /**
     * The ID of the banned user.
     */
    get userId() {
        return this[common_1.rawDataSymbol].user_id;
    }
    /**
     * The name of the banned user.
     */
    get userName() {
        return this[common_1.rawDataSymbol].user_login;
    }
    /**
     * The display name of the banned user.
     */
    get userDisplayName() {
        return this[common_1.rawDataSymbol].user_name;
    }
    /**
     * Retrieves more information about the user.
     */
    async getUser() {
        return (await this._client.users.getUserById(this[common_1.rawDataSymbol].user_id));
    }
    /**
     * The date when the ban will expire; null for permanent bans.
     */
    get expiryDate() {
        return this[common_1.rawDataSymbol].expires_at ? new Date(this[common_1.rawDataSymbol].expires_at) : null;
    }
};
tslib_1.__decorate([
    shared_utils_1.Enumerable(false)
], HelixBan.prototype, "_client", void 0);
HelixBan = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixBan', 'userId')
], HelixBan);
exports.HelixBan = HelixBan;
