// @twurple/api/lib/api/helix/schedule/HelixPaginatedScheduleSegmentRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixPaginatedScheduleSegmentRequest = void 0;
const common_1 = require("@twurple/common");
const HelixPaginatedRequest_1 = require("../HelixPaginatedRequest");
const HelixScheduleSegment_1 = require("./HelixScheduleSegment");
/**
 * A paginator specifically for schedule segments.
 */
class HelixPaginatedScheduleSegmentRequest extends HelixPaginatedRequest_1.HelixPaginatedRequest {
    /** @private */
    constructor(broadcaster, client, filter) {
        var _a;
        super({
            url: 'schedule',
            query: {
                broadcaster_id: common_1.extractUserId(broadcaster),
                start_time: filter === null || filter === void 0 ? void 0 : filter.startDate,
                utc_offset: (_a = filter === null || filter === void 0 ? void 0 : filter.utcOffset) === null || _a === void 0 ? void 0 : _a.toString()
            }
        }, client, data => new HelixScheduleSegment_1.HelixScheduleSegment(data, client), 25);
    }
    // sadly, this hack is necessary to work around the weird data model of schedules
    // while still keeping the pagination code as generic as possible
    /** @private */
    async _fetchData(additionalOptions = {}) {
        const origData = (await super._fetchData(additionalOptions));
        return {
            data: origData.data.segments,
            pagination: origData.pagination
        };
    }
}
exports.HelixPaginatedScheduleSegmentRequest = HelixPaginatedScheduleSegmentRequest;
