// @twurple/api/lib/api/helix/bits/HelixBitsApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixBitsApi = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@twurple/common");
const BaseApi_1 = require("../../BaseApi");
const HelixBitsLeaderboard_1 = require("./HelixBitsLeaderboard");
const HelixCheermoteList_1 = require("./HelixCheermoteList");
/**
 * The Helix API methods that deal with bits.
 *
 * Can be accessed using `client.bits` on an {@ApiClient} instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const leaderboard = await api.bits.getLeaderboard({ period: 'day' });
 * ```
 */
let HelixBitsApi = class HelixBitsApi extends BaseApi_1.BaseApi {
    /**
     * Retrieves a bits leaderboard of your channel.
     *
     * @expandParams
     */
    async getLeaderboard(params = {}) {
        const { count = 10, period = 'all', startDate, contextUserId } = params;
        const result = await this._client.callApi({
            type: 'helix',
            url: 'bits/leaderboard',
            scope: 'bits:read',
            query: {
                count: count.toString(),
                period,
                started_at: startDate ? startDate.toISOString() : undefined,
                user_id: contextUserId
            }
        });
        return new HelixBitsLeaderboard_1.HelixBitsLeaderboard(result, this._client);
    }
    /**
     * Retrieves all available cheermotes.
     *
     * @param broadcaster The broadcaster to include custom cheermotes of.
     *
     * If not given, only retrieves global cheermotes.
     */
    async getCheermotes(broadcaster) {
        const broadcasterId = broadcaster ? common_1.extractUserId(broadcaster) : undefined;
        const result = await this._client.callApi({
            type: 'helix',
            url: 'bits/cheermotes',
            query: {
                broadcaster_id: broadcasterId
            }
        });
        return new HelixCheermoteList_1.HelixCheermoteList(result.data);
    }
};
HelixBitsApi = tslib_1.__decorate([
    common_1.rtfm('api', 'HelixBitsApi')
], HelixBitsApi);
exports.HelixBitsApi = HelixBitsApi;
